compression/decompression-broken-chunks.html

```diff
@@ -0,0 +1,146 @@
+   ISIZE0 = compressedArrayWithGzip.length - 4,
+   ISIZE1 = compressedArrayWithGzip.length - 3,
+   ISIZE2 = compressedArrayWithGzip.length - 2,
+   ISIZE3 = compressedArrayWithGzip.length - 1];
+ const gzipByteoffsetNameList = [
+   'ID1', 'ID2', 'CM', 'FLG', 'MTIME', 'MTIME', 'MTIME', 'MTIME', 'XFL',
+   'OS', 'CRC32', 'CRC32', 'CRC32', 'CRC32', 'ISIZE', 'ISIZE', 'ISIZE', 'ISIZE'];
+
+ // First, make an incorrest input for by changing a |byte| of compressedArray
+ // indicated by the offset above into |byte| + |shift|. |shift| is from 1 to 255.
+ // Next, confirm that input is valid or invalid for DecompressionStream.
+ // If input is valid, decompress it and decode it to string (decompressedString).
+ // If decompressedString is correctly decompressed data, write like this:
+ //   [offset: 3, shift: 1 offset name: FLG]
+ // Else, write like this:
+ //   [offset: 17, shift: 7 offset name: DATA output: "expected outpupP"]
+
+ // Example: gzip, byteoffset: CM = 2, shift: 3, correct input: [31, 139, 8, 0, 0, ...]
+ // => incorrect input: [31, 139, 11, 0, 0, ...]
+
+ function write(s) {
+   let pre = document.querySelector('#results');
+   if (!pre) {
+     pre = document.createElement('pre');
+     pre.id = 'results';
+     document.body.appendChild(pre);
+   }
+   pre.textContent += s + '\n';
+ }
+
+ async function decompressArrayBuffer(input, format) {
+   const ds = new DecompressionStream(format);
+   const reader = ds.readable.getReader();
+   const writer = ds.writable.getWriter();
+   writer.write(input);
+   writer.close();
+   const out = [];
+   let totalSize = 0;
+   while (true) {
+     try {
+       const { value, done } = await reader.read();
+       console.log('read succeeded');
+       if (done) break;
+       out.push(value);
+       totalSize += value.byteLength;
+     } catch (e) {
+       console.assert(e.name == 'TypeError');
+       return 'invalid input';
+     }
+   }
+   const concatenated = new Uint8Array(totalSize);
+   let offset = 0;
+   for (const array of out) {
+     concatenated.set(array, offset);
+     offset += array.byteLength;
+   }
+   return concatenated;
+ }
+
+ async function decompressTest(format, input, byteoffsetList, byteoffsetNameList) {
+   write(`===== ${format} test =====`);
+   write(`correct compressed array of ${format}: ${input}`);
+   write(`expected output: 'expected output'`);
+   let correctList = [];
+   let incorrectList = [];
+   for (let offset = 0; offset < input.length; ++offset) {
+     for (let shift = 1; shift <= 255; ++shift) {
+       let brokenArray = input.slice();
+       brokenArray[offset] += shift;
+       const bufferView = new Uint8Array(brokenArray);
+       decompressedData = await decompressArrayBuffer(bufferView, format);
+       if (decompressedData !== 'invalid input') {
+         const decompressedString = new TextDecoder().decode(decompressedData);
+         let name = '';
+         if (byteoffsetList.includes(offset)) {
```

Comment:
Note: It's not important here, but the Array `includes()` method takes O(N) time. If you find yourself writing similar code in the future, think about using a hash table instead. For example 

```javascript
const deflateByteOffsetTable = {
  0: 'CMF',
  1: 'FLG',
  2: 'DICTID0',
  3: 'DICTID1',
  4: 'DICTID2',
  5: 'DICTID3',
  [compressedArrayWithDeflate.length - 4]: 'ADLER0',
  [compressedArrayWithDeflate.length - 3]: 'ADLER1',
  [compressedArrayWithDeflate.length - 2]: 'ADLER2',
  [compressedArrayWithDeflate.length - 1]: 'ADLER3',
};
```

Reply:
Thank you! I was looking for such a thing, but I couldn't,